[SE-0293] Extend property wrappers to function and closure parameters

hborla opened the pull request:
Implementation of [SE-0293: Extend property wrappers to function and closure parameters](https://github.com/apple/swift-evolution/blob/main/proposals/0293-extend-property-wrappers-to-function-and-closure-parameters.md).

Approach:
* Functions with property wrapper parameters use the backing property type in the function signature everywhere except the constraint system. When a function is resolved and its type is opened, the property wrapper type is "unwrapped" and replaced with the wrapped-value type or the projected-value type, depending on the argument label.
* When the property wrapper type is replaced, the property wrapper is "applied" by generating the appropriate constraints between the backing wrapper type and the wrapped or projected value type. The wrapper type and init kind (between wrapped-value and projected-value) are recorded in the constraint system.
* CSApply will wrap functions with wrapper parameters in an `autoclosure` that accepts the wrapped-value or projected-value type, which initializes the property wrapper and calls the underlying function. Other arguments that don't have property wrappers are simply forwarded.
  * Note: a lot of the code that builds the "forwarding" parameter list is duplicated in `buildCurryThunk`. I intend to clean this up.
  * If the function is immediately applied, no autoclosure is necessary and instead the property wrapper initialization expression is injected around the appropriate argument in `coerceCallArguments`.
  * Instead of injecting the initializer expression directly in CSApply, an `AppliedPropertyWrapperExpr` is used. When SILGen sees this expression, it will emit an application of the appropriate property wrapper generator function.
* When a function is emitted in SILGen, property-wrapped parameters are replaced with the backing property wrapper (which is synthesized as a `ParamDecl`). The local computed properties are emitted the same way as local property wrappers - by visiting the "auxiliary decls" during SILGen.
* When visiting parameter decls, SILGen will also emit both property wrapper generator functions if the parameter has them. If the enclosing function is public, property wrapper generator functions are serialized and given public non-ABI linkage.

A few things that are still a WIP:
* Unapplied references to functions with effects (e.g. `throws`, `async`)
* Diagnostics for invalid wrapper attributes in overrides and protocol witnesses
* "Implementation detail" property wrappers from the latest revision
* Closure parameter type inference
* Testing!

Not implemented in this PR:
* Nonmutating setter synthesis for inferred property wrappers (e.g. `$` closure parameters with no attribute)
* Property wrapper parameters in subscript declarations

Resolves: rdar://66866426, rdar://71902161

hborla committed:
[ConstraintSystem] Teach the constraint system about property wrapper
parameters.

hborla committed:
[Property Wrappers] Handle property wrappers on parameters in
PropertyWrapperBackingPropertyInfoRequest.

hborla committed:
[Property Wrappers] Don't allow property wrapper attributes to have
arguments when applied to parameters.

hborla committed:
[Property Wrappers] Add a request to synthesize the local wrapped value
variable for a parameter with an attached property wrapper.

hborla committed:
[Sema] Visit auxiliary decls on parameters in the decl checker.

hborla committed:
[SILGen] Teach SILGen to emit property wrapper parameters.

hborla committed:
[Property Wrappers] Allow property wrapper attributes to be applied
to parameters.

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please build toolchain macOS platform

hborla committed:
[ConstraintSystem] Use a wrapped value placeholder when matching property
wrapper arguments in the constraint system in order to avoid generating
constraints for the argument multiple times.

hborla commented:
@swift-ci please build toolchain

hborla committed:
[Property Wrappers] Set the interface type of the synthesized local
wrapped value var in CSApply.

This is necessary when closure parameter types are inferred.

hborla commented:
@swift-ci please build toolchain

hborla committed:
[Property Wrappers] Use the TypeExpr instead of the TypeRepr of the
property wrapper custom attribute to get the backing wrapper type
in CSApply.

This is necessary because implicit custom attributes do not have
TypeReprs, but they always have TypeExprs.

hborla committed:
[Property Wrappers] Determine the mutability of a property wrapper parameter
based on mutability of the property wrapper's wrappedValue setter.

hborla commented:
@swift-ci please build toolchain

hborla committed:
[Property Wrappers] Teach the constraint system to build init(projectedValue:)
calls.

hborla committed:
[PreCheckExpr] Strip $ prefixes out from compound decl names for unqualified
lookup in TypeChecker::resolveDeclRefExpr.

hborla committed:
[Property Wrappers] Don't include the $ prefix in synthesized property wrapper
var names when the original wrapped property represents a projected-value
closure parameter.

hborla committed:
[Parser] Allow $ prefixes on argument labels and closure parameter
declarations.

hborla committed:
[Property Wrappers] Don't allow property wrappers applied to parameters
to have a mutating wrappedValue getter.

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please build toolchain

hborla committed:
[Property Wrappers] Add a constraint fix and diagnostic for the case
where the programmer tries to pass a projection argument to a wrapped
parameter without 'var projectedValue'.

hborla committed:
[Sema] Diagnose invalid $ prefixes on closure parameters after the
closure is resolved. This can't be diagnosed earlier because a property
wrapper attribute may be inferred in the constraint system.

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please clean smoke test

hborla committed:
[ConstraintSystem] Emit an error message when trying to pass a property
wrapper projection argument to a wrapped parameter with arguments in the
wrapper attribute.

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please build toolchain

hborla committed:
[NFC][Property Wrappers] Start to add Sema tests for SE-0293.

hborla committed:
[Property Wrappers] Add an unsupported error message for property wrapper
parameters in subscript declarations.

hborla committed:
[Property Wrappers] Implement support for parameters with an implicit
property wrapper in the property wrapper requests.

hborla committed:
[ConstraintSystem] Implement implicit property wrapper attribute
inference for closure parameters.

hborla committed:
[NFC][Property Wrappers] Start to add SILGen tests for SE-0293.

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please build toolchain

hborla committed:
[CSGen] Move generateWrappedPropertyTypeConstraints back to its original
position in CSGen. This effectively reverts 6caca26014.

hborla committed:
[CSApply] Add buildPropertyWrapperFnThunk, which builds a thunk around
a reference to a function with property wrapper parameters and initializes
each backing property wrapper from a wrapped or projected value.

hborla committed:
[ASTWalker] Support configurating whether or not to visit the original
value of a property wrapper placeholder expression.

hborla committed:
[ConstraintSystem] Change the type of an unapplied function reference
when it has property wrapper parameters.

The property wrapper type will be replaced with either the wrapped-value
or projected-value type, depending on the argument label/parameter name,
and CSApply will build a thunk to construct the property wrapper and call
the function.

hborla committed:
[Property Wrappers] Store property wrapper "init from projection" expressions
in PropertyWrapperBackingPropertyInfo.

hborla committed:
[SILGen] Teach SILGen to emit property wrapper generator functions that
take in a projected value.

hborla committed:
[Sema] Add an implicit applied property wrapper expression and a new
dedicated initializer context for this expressions.

hborla committed:
[Property Wrappers] Compute generic wrapped-value and projected-value
initialization expressions for parameters with attached property wrappers.

hborla committed:
[ConstraintSystem] Inject implicit applied property wrapper expressions
around arguments to property wrapper parameters.

hborla commented on include/swift/AST/ASTWalker.h:
```diff
@@ -215,6 +215,10 @@ class ASTWalker {
 
+  /// This method configures whether the the walker should visit the underlying
+  /// value of a property wrapper placeholder.
+  virtual bool shouldWalkIntoPropertyWrapperPlaceholderValue() { return true; }
```
This is no longer used - I'll remove it

hborla commented on lib/Sema/CSApply.cpp:
```diff
@@ -3439,9 +3566,19 @@ namespace {
+      // expression cannot be reused, so we only need the original wrapped value
+      // argument in the AST.
+      if (!expr->getOpaqueValuePlaceholder())
+        return expr->getOriginalWrappedValue();
```
This is also no longer necessary

hborla commented:
I'm still implementing the [latest revision](https://forums.swift.org/t/pitch-5-se-0293-extend-property-wrappers-to-function-and-closure-parameters/45040) but most of the infrastructure is in place

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please build toolchain

hborla committed:
[SILGen] Serialize and give PublicNonABI linkage to property wrapper
generator functions for parameters of public functions.

hborla commented:
@swift-ci please build toolchain

hborla commented:
@swift-ci please build toolchain

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -1415,6 +1420,15 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
+                                             argLabel, subKind, locator).isFailure()) {
+          cs.getTypeMatchFailure(loc);
+        }
+        continue;
```
Should this also check the opposite case where argument has `$` but parameter doesn't have a property wrapper? If I understand code correctly label matching just looks over `$` so if the argument type matches parameter type in this case would calls like this be wellformed?

```swift
func test(v: Int) {}

var x: Int = 42

test($v: x)
```

hborla replied:
Hah, I think you're right. I didn't even think about that, thank you!

hborla replied:
✅

xedin reviewed (approved):
Looks good to me! I have left a couple of comments inline but nothing that should prevent this PR from being landed.

hborla commented on include/swift/AST/Expr.h:
```diff
@@ -4228,6 +4228,57 @@ class PropertyWrapperValuePlaceholderExpr : public Expr {
+  Expr *getValue() { return Value; }
+
+  /// Sets the value that the property wrapper is applied to.
+  void setValue(Expr *value) { Value = value; }
```
Yeah, this `setValue` is only used in the ASTWalker in case the value got rewritten. I don't think anything actually ever rewrites the expression though (but it does need to get visited)

xedin replied:
Sounds good, just making sure :)

hborla commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -1373,6 +1378,15 @@ ConstraintSystem::TypeMatchResult constraints::matchCallArguments(
+        auto argLabel = argInfo->Labels[argIdx];
+        if (cs.applyPropertyWrapperParameter(paramTy, argTy, const_cast<ParamDecl *>(param),
+                                             argLabel, subKind, locator).isFailure()) {
+          cs.getTypeMatchFailure(loc);
```
Fixed!

hborla committed:
[Property Wrappers] Store the callee in AppliedPropertyWrapperExpr to use
the correct substitution map when emitting the property wrapper generator
application in SILGen.

hborla committed:
[ConstraintSystem] Don't allow dollar prefixes in argument labels unless
the parameter has an attached property wrapper.

hborla committed:
[NFC][ConstraintSystem] Rename applyPropertyWrapperParameter.

hborla committed:
[Property Wrappers] Don't allow parameters with attached property wrappers
in protocol requirements.

hborla commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -4633,6 +4642,12 @@ class ConstraintSystem {
 
+  /// Matches a wrapped or projected value parameter type to its backing
+  /// property wrapper type by applying the property wrapper.
+  TypeMatchResult applyPropertyWrapperParameter(
```
✅

hborla commented:
@swift-ci please test

hborla commented:
@swift-ci please test source compatibility

hborla commented:
@swift-ci please smoke test

hborla commented:
@swift-ci please smoke test macOS platform

hborla commented:
@swift-ci please smoke test macOS platform

davezarzycki commented on lib/Sema/ConstraintSystem.cpp:
```diff
@@ -1631,6 +1690,16 @@ ConstraintSystem::getTypeOfMemberReference(
+    auto *functionType = fullFunctionType->getResult()->getAs<FunctionType>();
+    functionType = unwrapPropertyWrapperParameterTypes(*this, funcDecl, functionRefKind,
+                                                       functionType, locator);
+    openedType = FunctionType::get(fullFunctionType->getParams(), functionType);
```
Is there a reason why we're not doing any FunctionType::ExtInfo propagation/preservation here? In other words, are the default ExtInfo flags the goal or do they just happen to work?

And as a general observation that long predates your pull request, it's kind of frustrating that the FunctionType constructor let's one elide (and therefore default) the ExtInfo state. This wouldn't ordinarily be a problem for most constructors but Swift is constantly unpacking and rebuilding function types and I find that the ExtInfo preservation/management to be haphazard. I've been tempted to propose something but I don't know what people think/feel about `ExtInfo`.

hborla replied:
Ah, I think you're right that dropping them just happens to work here. Thanks for catching - I'll fix this in a followup PR.

Agreed that the default `ExtInfo()` makes it too easy to accidentally drop existing ExtInfo when transforming a function type like this. I actually had to fix the same mistake in the body of `unwrapPropertyWrapperParameterTypes` as well

davezarzycki replied:
Thanks for clarifying. A link to the followup PR would be appreciated. I just posted to the forum about making `ExtInfo` preservation/propagation more explicit:

https://forums.swift.org/t/rfc-make-extinfo-param-to-functiontype-construction-always-explicit/45182

davezarzycki replied:
Just to close the loop, I merged #36234 (but GitHub thinks the merge failed, weird). Thanks for helping me better understand the intent here and now that #36234 is merged, hopefully intent will be a little more clear in the future. :-)
